File: src/content/docs/topics/cs/coding-problems/binary-search/162-find-peak-element-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function findPeakElement(nums: number[]): number {
    let lo = 0, hi = nums.length - 1;
    while (lo < hi) {
        const mid = (lo + hi) >> 1;
        if (nums[mid] < nums[mid + 1]) lo = mid + 1;
        else hi = mid;
    }
    return lo;
}

const r1 = findPeakElement([1, 2, 3, 1]);
assert(r1 === 2, `expected 2, got ${r1}`);

const r2 = findPeakElement([1, 2, 1, 3, 5, 6, 4]);
assert(r2 === 1 || r2 === 5, `expected 1 or 5, got ${r2}`);

assert(findPeakElement([1]) === 0);
assert(findPeakElement([1, 2, 3]) === 2);
assert(findPeakElement([3, 2, 1]) === 0);
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/binary-search/162-find-peak-element-approach2.swift
// LEETCODE_TYPE: Solution

func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func findPeakElement(_ nums: [Int]) -> Int {
        var low = 0
        var high = nums.count - 1
        while low < high {
            let middle = low + (high - low) / 2
            if nums[middle] > nums[middle + 1] {
                high = middle
            } else {
                low = middle + 1
            }
        }
        return low
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:9cea99660cc25df19d165144a076beb7910c1b0838d6489b61fdf49ee8d7a218
    expectEqual(Solution().findPeakElement([1, 2, 3, 1]), 2, "interior-peak")
    expectEqual(Solution().findPeakElement([1, 3, 2]), 1, "short-interior-peak")
    expectEqual(Solution().findPeakElement([7]), 0, "single-element")
    expectEqual(Solution().findPeakElement([3, 2, 1]), 0, "strictly-decreasing")
    expectEqual(Solution().findPeakElement([1, 2, 3]), 2, "strictly-increasing")
    // EXCLUDED_VECTOR empty-array: [[]] | The problem contract requires at least one array element.
    // EXCLUDED_VECTOR equal-adjacent-values: [[1,1]] | Adjacent values must be different.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
